include/openthread/border_agent.h

```diff
@@ -123,6 +123,36 @@ bool otBorderAgentIsActive(otInstance *aInstance);
  */
 uint16_t otBorderAgentGetUdpPort(otInstance *aInstance);
 
+/**
+ * This callback informs the application the changes of the state of the MeshCoP service.
+ *
+ * In specific, the 'state' includes the MeshCoP TXT data originated from the Thread network and whether the
+ * Border Agent is Active (which can be obtained by `otBorderAgentIsActive`).
+ *
+ * @param[in] aTxtData  A pointer to the encoded MeshCoP TXT data originated from the Thread network.
+ * @param[in] aLength   The length of the encoded TXT data.
+ * @param[in] aContext  A pointer to application-specific context.
+ */
+typedef void (*otBorderAgentMeshCoPServiceChangedCallback)(const uint8_t *aTxtData, uint16_t aLength, void *aContext);
+
+/**
+ * Sets the callback function used by the Border Agent to notify any changes on the state of the MeshCoP service.
+ *
+ * The callback will be invoked when the 'Is Active' state of the Border Agent or the MeshCoP service TXT data values
+ * changes. For example, it will be invoked when the network name or the extended PAN ID changes and pass the updated
+ * encoded TXT data to the application layer.
+ *
+ * This callback will be invoked once right after this API is called to provide initial states of the MeshCoP service
+ * to the application.
+ *
+ * @param[in] aInstance  A pointer to an OpenThread instance.
+ * @param[in] aCallback  The callback to be invoked when there are any changes of the MeshCoP service.
+ * @param[in] aContext   A pointer to application-specific context.
+ */
+void otBorderAgentSetMeshCoPServiceChangedCallback(otInstance                                *aInstance,
+                                                   otBorderAgentMeshCoPServiceChangedCallback aCallback,
+                                                   void                                      *aContext);
+
```

Comment:
There is no implementation of these APIs in this PR. Do you plan to add them later in a separate PR? 

I think it would be better to add the API along with its implementation in the same PR to keep the codebase consistent.

Reply:
Sounds good! I had meant to let us agree on the API signature and usage first. I'll add the implementation in this PR together.

Reply:
I've added the implementation and the tests. It's a little bit long now. But most code is unit testing and Txt data encoding. The encoding process follows the process in ot-br-posix BorderAgent.